### ISSUE TITLE
INotebookTracker.panel.model.metadata.set fails to change notebook metadata

### ISSUE THREAD

**Issue by flying-sheep** [jupyterlab/jupyterlab#9183]
I use the following code in my extension to add an UUID to a notebook (It’s so short that I’ll just paste it whole):

```js
import { JupyterFrontEnd, JupyterFrontEndPlugin } from '@jupyterlab/application'
import { INotebookTracker, NotebookPanel } from '@jupyterlab/notebook'
import { UUID } from '@lumino/coreutils'

const extension: JupyterFrontEndPlugin<void> = {
  id: 'notebook-uuid',
  autoStart: true,
  requires: [INotebookTracker],
  activate: (app: JupyterFrontEnd, notebooks: INotebookTracker) => {
    notebooks.forEach(addUUID)
    notebooks.widgetAdded.connect((_, a) => addUUID(a))
  },
}

function addUUID(panel: NotebookPanel): void {
  if (panel.model === null) {return}
  const { metadata } = panel.model
  if (metadata.has('uuid')) {return}
  metadata.set('uuid', UUID.uuid4())
  console.log(`Added UUID ${metadata.get('uuid')} to notebook`)
}

export default extension
```

This code runs fine, and doesn’t return early. (It logs the message at the end)

It also *used* to actually add UUIDs to the notebook metadata, but no change is displayed in the “Notebook metadata” view or the notebook file’s JSON. Jupyter lab on master seems to use that object the same way, but what it does seems to work:

https://github.com/jupyterlab/jupyterlab/blob/011ff58a61689b734e29a2fe2c97c9aaa52efe25/packages/notebook/src/panel.ts#L225

I use Jupyter lab 2.2.8

Asked in Gitter here: https://gitter.im/jupyterlab/jupyterlab?at=5f9151cdeb82301c1a402556

**Comment by flying-sheep** [comment]
This is absolutely wild! Note how it says Map(3) on the left, but Map(2) in the debugger:

![grafik](https://user-images.githubusercontent.com/291575/98541442-c07cbc80-228f-11eb-85ad-96cdf128a0b6.png)

Does something reset it after the extension runs?

**Comment by agoose77** [comment]
I have replied on Gitter, but for posterity, I copy my reply below:

 I think you need to wait until the context is ready:
```
function addUUID(panel: NotebookPanel): void {
    panel.context.ready.then(() => {

          const { metadata } = panel.model
          if (metadata.has('uuid')) {return}
          metadata.set('uuid', UUID.uuid4())
          console.log(`Added UUID ${metadata.get('uuid')} to notebook`)
          panel.context.save();
    });
}```

**Comment by flying-sheep** [comment]
Thank you, it works! I would never have figured this out on my own. Why descend into some arbitrary variable and wait on some property of it?

I’m confused: The API seems to take great care that nothing can be accessed without it being available:

- No global variables
- Dependency injection
- Separation of interfaces and implementations
- …

But this is just some magic invocation nobody can anticipate.

Can things be changed so a loud, clear error is thrown telling people to do the magic dance?

**Comment by agoose77** [comment]
So, I'm not a very experienced developer when it comes to the JupyterLab code base, but I assume it is related to the fact that the notebook model is synchronised with the backend (modelDB). Before the backend is connected, I wonder whether updates are not queued. 

More experienced developers might have a more concrete understanding.

I understand why this bug was frustrating - it seems that "undefined" (it may be defined as forbidden somewhere, but I haven't seen it) behaviour changed, and there is no clear overarching documentation to tell you what you *should* be doing.

To explain how I approached the issue, I started by assuming that there must be some context that needs to become ready; there are other places in the JupyterLab codebase where you need to wait for things to become available. Also, I was previously aware that the logic that reads/writes from disk is in the server extension, so it was possibly a bug relating to the communication with the backend. Initially I checked that you could modify the metadata from the save handler, and from there looked for some kind of "ready" attribute on the context object. 

I'll wait for someone more experienced to drop in here, but once they do, perhaps we can update the documentation so that people don't run in to this again!

**Comment by flying-sheep** [comment]
> and there is no clear overarching documentation to tell you what you should be doing

That wouldn’t be enough. The API is well-designed enough that I was 100% sure this is a bug and not intentional behavior. This kind of care in design doesn’t fit together with undefined behavior.

I was sure that if I can access an interface, I’m allowed to use the interface, because there’s multiple places in my tiny codebase where I *have* to register a callback or dependency before the type system allows me to access something.

If everything was hacky globals, I’d have sighed and searched for some magic thing I need to `await` somewhere.

> I wonder whether updates are not queued

Hmm, that would probably work in my case, as I don’t rely on the change taking effect: I “fire and forget”. Queuing changes would mean that many many methods would become asynchronous, a huge API change.

----

In short: The API is a weird mishmash of undefined behavior and beautiful avoidance of the same. I don’t want to be able to access a non-promise `.metadata` that could be silently unusable. How can we fix the API so this is no longer possible?

**Comment by agoose77** [comment]
> Hmm, that would probably work in my case, as I don’t rely on the change taking effect: I “fire and forget”. Queuing changes would mean that many many methods would become asynchronous, a huge API change.

Not in this case; the update queue would be an internal detail that was read once the backend is connected read/write methods would remain synchronous.

As to your comments on the design, I think it really just depends upon whether there are use cases for obtaining the model instance before it has been loaded. My initial guess is that the UI for the notebook is created before it has all been loaded, which makes sense (to me) for UI responsiveness. In this case, it would be an improvement if model changes were synchronised to the backend once available.

**Comment by flying-sheep** [comment]
> Not in this case; the update queue would be an internal detail that was read once the backend is connected read/write methods would remain synchronous.

Ah, so there’d be a ephemeral map backend that you can modify, and the context flushes all changes once it’s reday. That could work!

> I think it really just depends upon whether there are use cases for obtaining the model instance before it has been loaded.

Well, if there’s e.g. some initial readonly `metadata` that could be useful, it should be a `ImmutableSet` without modifying methods. The mutable version that relies on some promise being resolved should itself sit behind a promise.

So I assume a better design here would have been `initialMetadata: IImmutableJSON, metadata: IObservableJSON` or so
